### DISCUSSION TITLE
关于C# 泛型约束的建议

### DISCUSSION THREAD

**Discussion by fanyu0122** [dotnet/csharplang#6075] [Language Ideas]
希望C#的泛型约束可以用户自定义，比如定义这个泛型必须被某个自定义Attribute标记

**Comment by huoyaoyuan** [comment]
Translation: generic constraint should be definable by user, for example require some custom Attribute.

**Comment by huoyaoyuan** [comment]
Answer in short: the C# team are generally not interested in constraints to restrict usages only. That can be done by an analyzer reporting error. Current constraints are enabling specific features requiring the constraint, for example calling interface methods.

简要回答：C#团队对仅仅做出限制的约束语法并不感兴趣。可以使用自定义分析器来对需要的限制进行报错。目前的泛型约束都是在约束之上开启了新的功能，比如调用接口中的方法。

**Comment by TahirAhmadov** [comment]
You can define a static constructor which will verify that generic argument(s) satisfies these "custom" constraints. Then, it will fail at runtime, but very early and with a good error message that lets you find the issue quickly.
